cmd/crank/xpkg/xpextract.go

```diff
@@ -0,0 +1,218 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package xpkg
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/spf13/afero"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/internal/xpkg"
+	"github.com/crossplane/crossplane/internal/xpkg/upbound"
+)
+
+const (
+	errMustProvideTag          = "must provide package tag if fetching from registry or daemon"
+	errInvalidTag              = "package tag is not a valid reference"
+	errFetchPackage            = "failed to fetch package from remote"
+	errGetManifest             = "failed to get package image manifest from remote"
+	errFetchLayer              = "failed to fetch annotated base layer from remote"
+	errGetUncompressed         = "failed to get uncompressed contents from layer"
+	errMultipleAnnotatedLayers = "package is invalid due to multiple annotated base layers"
+	errOpenPackageStream       = "failed to open package stream file"
+	errCreateOutputFile        = "failed to create output file"
+	errCreateGzipWriter        = "failed to create gzip writer"
+	errExtractPackageContents  = "failed to extract package contents"
+)
+
+const (
+	layerAnnotation     = "io.crossplane.xpkg"
+	baseAnnotationValue = "base"
+	cacheContentExt     = ".gz"
+)
+
+// fetchFn fetches a package from a source.
+type fetchFn func(context.Context, name.Reference) (v1.Image, error)
+
+// registryFetch fetches a package from the registry.
+func registryFetch(ctx context.Context, r name.Reference) (v1.Image, error) {
+	return remote.Image(r, remote.WithContext(ctx))
+}
+
+// daemonFetch fetches a package from the Docker daemon.
+func daemonFetch(ctx context.Context, r name.Reference) (v1.Image, error) {
+	return daemon.Image(r, daemon.WithContext(ctx))
+}
+
+func xpkgFetch(path string) fetchFn {
+	return func(ctx context.Context, r name.Reference) (v1.Image, error) {
+		return tarball.ImageFromPath(filepath.Clean(path), nil)
+	}
+}
+
+// AfterApply constructs and binds Upbound-specific context to any subcommands
```

Comment:
It seems odd to have Upbound-specific code in Crossplane core - maybe this got ported and not renamed so it's not really Upbound-specific anymore?  We should probably refactor the whole upbound module into something that is not-upbound.

Reply:
https://github.com/crossplane/crossplane/tree/main/internal/xpkg/upbound but this is already in main - so would be great to handle this in a seperate PR

Reply:
Definitely, but maybe reword the comment so it doesn't imply that the context is actually upbound-specific?  I'll open an issue for the refactoring.

Reply:
done